File: frontend/src/types/role.ts
export interface Role {
  id: string;
  tenant_id: string;
  name: string;
  permissions: Record<string, string[]>;
  is_system_role: boolean;
  created_at: string;
  updated_at: string;
}

export interface RoleCreate {
  name: string;
  permissions: Record<string, string[]>;
  is_system_role?: boolean;
}

export interface RoleUpdate {
  name?: string;
  permissions?: Record<string, string[]>;
}

File: backend/app/models/role.py
from pydantic import BaseModel, Field
from typing import Dict, Any, Optional
from datetime import datetime
from uuid import UUID


class RoleBase(BaseModel):
    name: str
    permissions: Dict[str, Any] = Field(default_factory=dict, description="Permission matrix as JSON")
    is_system_role: bool = False


class RoleCreate(RoleBase):
    tenant_id: UUID


class RoleUpdate(BaseModel):
    name: Optional[str] = None
    permissions: Optional[Dict[str, Any]] = None


class Role(RoleBase):
    id: UUID
    tenant_id: UUID
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PermissionBase(BaseModel):
    resource: str
    action: str
    description: Optional[str] = None


class Permission(PermissionBase):
    id: UUID
    created_at: datetime

    class Config:
        from_attributes = True

